File: python/chirp.py
"""
Chirp Generator — LFM (Linear Frequency Modulation) waveform generation.

Generates complex baseband chirp signals for pulse compression radar.
"""

import numpy as np
import matplotlib.pyplot as plt

# Default parameters
DEFAULT_SAMPLE_RATE = 200_000      # Hz
DEFAULT_NUM_SAMPLES = 64
DEFAULT_F_START = 39_000           # Hz
DEFAULT_F_END = 41_000             # Hz


def generate_chirp(f_start, f_end, num_samples, sample_rate):
    
    """
    Generate a complex baseband LFM chirp signal.

    Args:
        f_start:     Start frequency of the sweep (Hz)
        f_end:       End frequency of the sweep (Hz)
        num_samples: Number of samples in the chirp
        sample_rate: Sample rate (Hz)

    Returns:
        Complex numpy array of shape (num_samples,) containing I + jQ samples.
    """
    t = np.arange(num_samples) / sample_rate
    
    # Chirp frequency sweep: f(t) = f_start + (f_end - f_start) * t / T
    # where T is the total time duration (num_samples / sample_rate)
    T = num_samples / sample_rate
    f_t = f_start + (f_end - f_start) * t / T
    
    # Phase accumulation: phi(t) = 2 * pi * integral(f(t) dt)
    # The integral of f(t) is f_start*t + 0.5 * (f_end - f_start)/T * t^2
    phase = 2 * np.pi * (f_start * t + 0.5 * (f_end - f_start) / T * t**2)
    
    # Complex signal: s(t) = exp(j * phase)
    chirp_signal = np.exp(1j * phase)
    return chirp_signal


def plot_chirp(chirp_signal, sample_rate):
    """
    Plot the chirp: real part, imaginary part, and spectrogram.

    Use this to visually verify that the instantaneous frequency
    sweeps linearly from f_start to f_end.

    Args:
        chirp_signal: Complex numpy array from generate_chirp()
        sample_rate:  Sample rate (Hz)
    """
    plt.figure()
    plt.subplot(3, 1, 1)
    plt.plot(chirp_signal.real)
    plt.title("Real Part")
    plt.xlabel("Sample")
    plt.ylabel("Amplitude")

    plt.subplot(3, 1, 2)
    plt.plot(chirp_signal.imag)
    plt.title("Imaginary Part")
    plt.xlabel("Sample")
    plt.ylabel("Amplitude")

    plt.subplot(3, 1, 3)
    plt.specgram(chirp_signal, Fs=sample_rate)
    plt.title("Spectrogram")
    plt.xlabel("Time (s)")
    plt.ylabel("Frequency (Hz)")

    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    chirp = generate_chirp(DEFAULT_F_START, DEFAULT_F_END,
                           DEFAULT_NUM_SAMPLES, DEFAULT_SAMPLE_RATE)
    plot_chirp(chirp, DEFAULT_SAMPLE_RATE)

File: java/Chirp.java
/*
 * ===============================================================================
 * Chirp — Linear FM chirp generator (floating point reference).
 * ===============================================================================
 *
 * Produces a complex LFM chirp as two parallel arrays (real, imaginary).
 * This is the floating-point "truth" that the VHDL DDS chirp generator
 * will be compared against.
 * ===============================================================================
 */
public class Chirp {

    /**
     * Generate a complex LFM (Linear Frequency Modulation) chirp.
     *
     * A chirp is a sinusoid whose frequency changes over time. For an LFM
     * chirp, the frequency increases (or decreases) linearly from startFreq
     * to endFreq over the duration of the pulse.
     *
     * The math:
     *
     *   Instantaneous frequency:   f(t) = f_start + (f_end - f_start) * t / T
     *     where T is the pulse duration (num_samples / sample_rate).
     *
     *   Phase is the integral of frequency:
     *     phase(t) = 2π * ∫ f(τ) dτ
     *              = 2π * (f_start * t + (f_end - f_start)/(2T) * t²)
     *
     *   Complex baseband signal:
     *     s(t) = exp(j * phase(t)) = cos(phase(t)) + j*sin(phase(t))
     *
     * Why complex? A real sinusoid carries both positive AND negative
     * frequency components that mirror each other. A complex exponential
     * carries only one frequency, which simplifies matched filtering and
     * makes the math cleaner. In hardware, the complex signal is implemented
     * as two real signal paths (I = cos, Q = sin) running in parallel.
     *
     * Why a linear sweep specifically? The autocorrelation of an LFM chirp
     * has a very sharp central peak, which is exactly what pulse compression
     * needs. Other waveforms exist (Barker codes, polyphase codes, nonlinear
     * chirps) but LFM is the workhorse because it's easy to generate in
     * hardware and gives great compression for its bandwidth.
     *
     * @param startFrequencyHz Frequency at t=0
     * @param endFrequencyHz   Frequency at t=T (pulse end)
     * @param numSamples       Chirp length in samples
     * @param sampleRate       Sample rate in Hz (determines T)
     * @return Two parallel arrays: [realSamples, imaginarySamples]
     */
    public static double[][] generate(double startFrequencyHz,
                                      double endFrequencyHz,
                                      int numSamples,
                                      int sampleRate) {
        double[] realSamples      = new double[numSamples];
        double[] imaginarySamples = new double[numSamples];

        // Pulse duration in seconds: T = N / Fs.
        double chirpDurationSeconds = (double) numSamples / sampleRate;

        for (int sampleIndex = 0; sampleIndex < numSamples; sampleIndex++) {
            // Sample time in seconds: t = n / Fs.
            double timeSeconds = (double) sampleIndex / sampleRate;

            // Quadratic phase formula:
            //   phase(t) = 2π * (f_start * t + (f_end - f_start) / (2T) * t²)
            //
            // The first term (f_start * t) is a linear phase from the
            // starting frequency. The second term (the t² term) is what
            // makes the chirp sweep — its derivative with respect to time
            // is (f_end - f_start)/T, which is exactly the rate at which
            // frequency changes per second.
            double instantaneousPhase = 2.0 * Math.PI
                    * (startFrequencyHz * timeSeconds
                      + 0.5 * (endFrequencyHz - startFrequencyHz)
                            / chirpDurationSeconds
                            * timeSeconds * timeSeconds);

            // Complex exponential: exp(j*phase) = cos(phase) + j*sin(phase).
            realSamples[sampleIndex]      = Math.cos(instantaneousPhase);
            imaginarySamples[sampleIndex] = Math.sin(instantaneousPhase);
        }
        return new double[][] { realSamples, imaginarySamples };
    }
}
